src/Features/Core/Portable/MetadataAsSource/MetadataAsSourceHelpers.cs

```diff
@@ -96,28 +103,244 @@ public static async Task<Location> GetLocationInGeneratedSourceAsync(SymbolKey s
+        /// <summary>
+        /// Uses various heuristics to try to find the implementation assembly for a reference assembly without
+        /// loading 
+        /// </summary>
+        public static bool TryFindImplementationAssemblyPath(string referencedDllPath, [NotNullWhen(true)] out string? implementationDllPath)
         {
+            if (TryNugetLibToRef(referencedDllPath, out implementationDllPath))
+                return true;
+
+            if (TryTargetingPackToSharedSdk(referencedDllPath, out implementationDllPath))
+                return true;
+
             implementationDllPath = null;
+            return false;
+
+            static bool TryNugetLibToRef(string referenceDllPath, [NotNullWhen(true)] out string? implementationDllPath)
+            {
+                implementationDllPath = null;
+
+                // For some nuget packages if the reference path has a "ref" folder in it, then the implementation assembly
+                // will be in the corresponding "lib" folder.
+                var start = referenceDllPath.IndexOf(@"\ref\");
+                if (start == -1)
+                    return false;
+
+                var pathToTry = referenceDllPath.Substring(0, start) +
+                                @"\lib\" +
+                                referenceDllPath.Substring(start + 5);
+
+                if (IOUtilities.PerformIO(() => File.Exists(pathToTry)))
+                {
+                    implementationDllPath = pathToTry;
+                    return true;
+                }
 
-            // For some nuget packages if the reference path has a "ref" folder in it, then the implementation assembly
-            // will be in the corresponding "lib" folder.
-            // TODO: Support more cases, like SDK references: https://github.com/dotnet/sdk/issues/12360
-            var start = referenceDllPath.IndexOf(@"\ref\");
-            if (start == -1)
                 return false;
+            }
+
+            static bool TryTargetingPackToSharedSdk(string referencedDllPath, [NotNullWhen(true)] out string? implementationDllPath)
+            {
+                implementationDllPath = null;
+
+                // eg. C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.5\ref\net6.0\Foo.dll
+                var parts = referencedDllPath.Split(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
+                if (parts is not [.., "packs", var packName, var packVersion, "ref", _, var dllFileName])
+                    return false;
+
+                var basePath = referencedDllPath.Substring(0, referencedDllPath.IndexOf("packs"));
+
+                // We try to get the shared sdk name from the FrameworkList.xml file, in the data dir
+                // eg. C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.5\data\FrameworkList.xml
+                var frameworkXml = Path.Combine(basePath, "packs", packName, packVersion, "data", "FrameworkList.xml");
```

Comment:
>basePath [](http://example.com/codeflow?start=48&length=8)

better: `Path.Combine(Path.GetDirectoryName(referenceDllPath), "..", "..", "FrameworkList.xml")`

Reply:
lol, this is what I orignally did for `basePath`, but six `"..", ` in a row looked really ugly and I figured someone would complain 😛